[Chore] Separate out attention backend constants from vllm.utils

hezyin opened the pull request:
## Purpose

Part of https://github.com/vllm-project/vllm/issues/26900

Separate out attention backend constants from the monolithic `vllm/utils/__init__.py` into a dedicated `vllm/attention/constants.py` module.

The following constants have been moved:
- `vllm.utils.STR_BACKEND_ENV_VAR` -> `vllm.attention.constants.STR_BACKEND_ENV_VAR`
- `vllm.utils.STR_FLASHINFER_ATTN_VAL` -> `vllm.attention.constants.STR_FLASHINFER_ATTN_VAL`
- `vllm.utils.STR_TORCH_SDPA_ATTN_VAL` -> `vllm.attention.constants.STR_TORCH_SDPA_ATTN_VAL`
- `vllm.utils.STR_XFORMERS_ATTN_VAL` -> `vllm.attention.constants.STR_XFORMERS_ATTN_VAL`
- `vllm.utils.STR_FLASH_ATTN_VAL` -> `vllm.attention.constants.STR_FLASH_ATTN_VAL`
- `vllm.utils.STR_INVALID_VAL` -> `vllm.attention.constants.STR_INVALID_VAL`

Backward compatibility is maintained through deprecation warnings.

## Test Plan

Run existing attention-related tests:
```bash
pytest tests/kernels/attention/test_attention_selector.py
pytest tests/kernels/attention/test_rocm_attention_selector.py
pytest tests/models/quantization/test_fp8.py
```

## Test Result

- Verified Python syntax is valid for all modified files
- Verified import paths are correct
- Backward compatibility maintained via `__getattr__` mechanism
- CI tests will validate full compatibility

hezyin committed:
[Chore] Separate out attention backend constants from vllm.utils

Part of https://github.com/vllm-project/vllm/issues/26900

This PR extracts attention backend-related constants from the monolithic
vllm/utils/__init__.py into a dedicated vllm/attention/constants.py module.

Changes:
- Created vllm/attention/constants.py with all attention backend constants
- Updated vllm/utils/__init__.py with deprecation warnings for backward compatibility
- Updated imports in vllm/attention/selector.py to use the new module
- Updated test files to import from the new location:
  - tests/kernels/attention/test_attention_selector.py
  - tests/kernels/attention/test_rocm_attention_selector.py
  - tests/models/quantization/test_fp8.py

The following constants have been moved:
- vllm.utils.STR_BACKEND_ENV_VAR -> vllm.attention.constants.STR_BACKEND_ENV_VAR
- vllm.utils.STR_FLASHINFER_ATTN_VAL -> vllm.attention.constants.STR_FLASHINFER_ATTN_VAL
- vllm.utils.STR_TORCH_SDPA_ATTN_VAL -> vllm.attention.constants.STR_TORCH_SDPA_ATTN_VAL
- vllm.utils.STR_XFORMERS_ATTN_VAL -> vllm.attention.constants.STR_XFORMERS_ATTN_VAL
- vllm.utils.STR_FLASH_ATTN_VAL -> vllm.attention.constants.STR_FLASH_ATTN_VAL
- vllm.utils.STR_INVALID_VAL -> vllm.attention.constants.STR_INVALID_VAL

Backward compatibility is maintained through deprecation warnings.

Signed-off-by: Henry Yin <yinhezheng1995@gmail.com>

hezyin committed:
Fix ruff formatting for import statement

Apply ruff's multi-line import formatting style as required by
pre-commit checks.

Signed-off-by: Henry Yin <yinhezheng1995@gmail.com>

hezyin committed:
Fix backward compatibility for constants outside vllm.utils

The __getattr__ function in vllm/utils/__init__.py was hardcoded to only
handle deprecated imports within vllm.utils (e.g., vllm.utils.profiling).
This broke backward compatibility for attention constants moved to
vllm.attention.constants, as it tried to import from the non-existent
path vllm.utils.attention.constants.

Changes:
- Detect when deprecated mapping points outside vllm.utils (contains ".")
- Construct correct import path: vllm.<module> vs vllm.utils.<module>
- Update deprecation warning to show correct new location
- Maintain backward compatibility for both cases:
  * Within vllm.utils: vllm.utils.profiling.cprofile
  * Outside vllm.utils: vllm.attention.constants.STR_BACKEND_ENV_VAR

This ensures old imports like "from vllm.utils import STR_BACKEND_ENV_VAR"
continue to work with proper deprecation warnings.

Fixes issue flagged by gemini-code-assist review.

Signed-off-by: Henry Yin <yinhezheng1995@gmail.com>

DarkLight1337 reviewed (commented):
Can you temporarily comment out the back compatibility code? To make sure that the deprecation warning is not triggered by internal code, I will run the tests to ensure that vLLM code has been updated. Afterwards, you can uncomment the code so downstream code still gets back-compatibility.

hezyin committed:
Temporarily disable backward compatibility for testing

Per reviewer request (DarkLight1337), temporarily commenting out the
deprecated mappings for attention backend constants to verify that all
internal vLLM code has been updated to use the new import paths from
vllm.attention.constants.

This ensures:
- No internal vLLM code triggers deprecation warnings
- All internal imports use the new location
- Tests will fail if any internal code still uses old imports

Once tests pass, the mappings will be uncommented to provide backward
compatibility for downstream code (e.g., lm-eval).

See: https://github.com/vllm-project/vllm/pull/28066#discussion_r1869045000
Signed-off-by: Henry Yin <yinhezheng1995@gmail.com>

hezyin committed:
Update remaining internal imports and re-enable backward compatibility

- Fixed tests/kernels/utils.py to import from vllm.attention.constants
- Re-enabled backward compatibility mappings in vllm/utils/__init__.py
- All internal vLLM code now uses the new import paths

The mypy error in tests/kernels/utils.py:1311 is pre-existing and unrelated
to this refactoring (our changes are on lines 19-23).

Signed-off-by: Henry Yin <yinhezheng1995@gmail.com>

DarkLight1337 commented:
I prefer Option 2. We should put the constants in a separate file just like with `mem_constants`
